## C-InterSecture
**C-InterSecture** (*C*omputional tool for *InterS*pecies analysis of genome archit*ecture*) pipline is python 2.7 based utilits to cross-species comparison of Hi-C map. C-InterSecture was designed to liftover contacts between species, compare 3-dimensional organization of defined genomic regions, such as TADs, and analyze statistically of individual contact frequencies.
 
# Prerequisites:
- numpy >= 1.9.0
- scipy >= 1.1.0
- [genome module](https://mirnylab.bitbucket.io/hiclib/_modules/mirnylib/genome.html) from mirnylib package with dependencies (it's minimal required module is included)

# Installation
There is no need for installation.

# Using C-InterSecture
C-InterSecture pipeline involves three step: data preprocessing, contact liftovering and, finally, visualization. Scripts for each step are placed in special folder. 

### Data preprocessing
The data preprocessing includes contacts filtration, statistical analysis and equalization. This step requires genome fasta-files, bin coordinates, raw and normalized contact matrices outputed by HiC-Pro. The corresponding scripts are located in "0_preprocessing" folder.
The preproccessing requires two annotation files: the chromosome order file and N-bases bin coverage. The chromosome order file contains a list of chromosomes with any additional information (for example, chromosome sizes):
```
chr1 something
chr2 something
...
```
THe N-bases bin coverage files are generated by the `unmappedBins.py` utility taking genome fasta-file. 
```
python unmappedBins.py < unmap.ini
```
The file `unmapped.ini` includes a space/tab delimited list of genome fasta files and bin sizes:
```
# any comments
path_to_sp1.fa resolution1 resolution2
path_to_sp2.fa resolution1 resolution2 resolution3
...
```
Output files are (and must been) placed with genome file and named as `geneme_file_name.resolution.unmap`. The N-bases bin coverage are stored as bed-liked format:
```
chrName bin_start bin_end N-bases coverage
```
After generation `.unmap`-files, can run preprocessing
```
python pre.py < pre.ini
```
Ini-file contains all neaded paramaters to performing:

**genome** - a path to genome fast file, folder with it must contains `.unmap`-files generated for given bin_sizes
**chrom_order** - a path to file contained chromosome order
**sample_name** - a name of sample, it are used to generate out-files names.
**genome_bins** = /mnt/scratch/ws/manuriddinov/201812071119EVL6/bcm/Ver/IMR90/raw/40000/IMR90_40000_abs.bed # for HiC-Pro output, genome coordinate of bins
raw_contacts = /mnt/scratch/ws/manuriddinov/201812071119EVL6/bcm/Ver/IMR90/raw/40000/IMR90_40000.matrix # path to raw matrix
norm_contacts = /mnt/scratch/ws/manuriddinov/201812071119EVL6/bcm/Ver/IMR90/iced/40000/IMR90_40000_iced.matrix # path to normalized matrix
statistic = abs # caluclated statistic, ether "abs" for absolute contacts value or "prc" for percentiled contacts
resolution = 40000 # hi-c map resolution, bin size of data
unmapped_bases = 33 # % N-bases into bin, if %>threshold bin discarded
coverage = 1 # discarding bins with coverage equal or less then bottom % bins
max_distance = 5 # maximal distance in MB for calculated individual distance-dependend statistic (other contacts are going to be united) 
out_folder = /mnt/scratch/ws/manuriddinov/201812071119EVL6/ArChER/pre $ folder to output

